File: 0139-word-break/0139-word-break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
  const set = new Set(wordDict);
  const ok = [true];
  
  for (let i = 1; i <= s.length; i++) {
    ok.push(check(s, i, set, ok));
  }
  
  return ok.at(-1);
};

const check = (s, i, set, ok) => {
  for (let j = 0; j < i; j++) {
    if (set.has(s.slice(j, i)) && ok[j]) return true;
  }
  
  return false;
}

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        seen = set(wordDict)
        ok = [True]
        
        for i in range(1, len(s) + 1):
            curr = False
            for j in range(0, i):
                if s[j:i] in seen and ok[j]:
                    curr = True
            ok.append(curr)
        return ok[-1]